### PR TITLE
Add evm precompile api

### PR DESCRIPTION
**Pull request by zjb0807**
Closes: #586

### CODE REVIEW

File: runtime/acala/src/lib.rs [hunk 1]
```diff
@@ -1250,6 +1250,7 @@ pub type MultiCurrencyPrecompile =
 	runtime_common::MultiCurrencyPrecompile<AccountId, EvmAddressMapping<Runtime>, Currencies>;
 
 pub type NFTPrecompile = runtime_common::NFTPrecompile<AccountId, EvmAddressMapping<Runtime>, NFT>;
+pub type EVMPrecompile = runtime_common::EVMPrecompile<AccountId, EvmAddressMapping<Runtime>, EVM>;
```
Comment (shaopengw) [inline runtime/acala/src/lib.rs:1253]: Rename it to `StateRentPrecompile`

File: runtime/common/src/precompile/evm.rs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+///
+/// `input` data starts with `action`.
+///
+/// Actions:
+/// - QueryContractExistentialDeposit.
+/// - QueryTransferMaintainerDeposit.
+/// - QueryStorageDepositPerByte.
+/// - QueryStorageDefaultQuota.
+/// - QueryMaintainer.
+/// - AddStorageQuota. Rest `input`bytes: `from`, `contract`, `bytes`.
```
Comment (shaopengw) [inline runtime/common/src/precompile/evm.rs:20]:
```suggestion
/// - AddStorageQuota. Rest `input` bytes: `from`, `contract`, `bytes`.
```

Comment (shaopengw) [reply runtime/common/src/precompile/evm.rs:20]: Same to rest documentations.

File: runtime/common/src/precompile/evm.rs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+/// - RemoveStorageQuota. Rest `input`bytes: `from`, `contract`, `bytes`.
+/// - RequestTransferMaintainer. Rest `input`bytes: `from`, `contract`.
+/// - CancelTransferMaintainer. Rest `input`bytes: `from`, `contract`.
+/// - ConfirmTransferMaintainer. Rest `input`bytes: `from`, `contract`,
+///   `new_maintainer`.
+/// - RejectTransferMaintainer. Rest `input`bytes: `from`, `contract`,
+///   `invalid_maintainer`.
+pub struct EVMPrecompile<AccountId, AddressMapping, EVM>(PhantomData<(AccountId, AddressMapping, EVM)>);
```
Comment (shaopengw) [inline runtime/common/src/precompile/evm.rs:28]:
```suggestion
pub struct StateRentPrecompile<AccountId, AddressMapping, EVM>(PhantomData<(AccountId, AddressMapping, EVM)>);
```

File: runtime/common/src/precompile/evm.rs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+			Action::QueryStorageDefaultQuota => {
+				let quota = vec_u8_from_u32(EVM::query_storage_default_quota());
+				Ok((ExitSucceed::Returned, quota, 0))
+			}
+			Action::QueryMaintainer => {
+				let contract = input.evm_address_at(1)?;
+
+				EVM::query_maintainer(contract).map_err(|e| ExitError::Other(Cow::Borrowed(e.into())))?;
+
+				Ok((ExitSucceed::Returned, vec![], 0))
```
Comment (xlc) [inline runtime/common/src/precompile/evm.rs:101]: where is the returned result?

**Comment by xlc** [comment]
I think we need more tests

File: runtime/common/src/precompile/mod.rs [hunk 5]
```diff
@@ -11,10 +11,12 @@ use primitives::PRECOMPILE_ADDRESS_START;
 use sp_core::H160;
 use sp_std::{marker::PhantomData, prelude::*};
 
+pub mod evm;
```
Comment (shaopengw) [inline runtime/common/src/precompile/mod.rs:14]:
```suggestion
pub mod state_rent;
```

File: runtime/common/src/precompile/mod.rs [hunk 6]
```diff
@@ -25,15 +27,21 @@ pub type EthereumPrecompiles = (
...
+		PrecompileCallerFilter,
+		MultiCurrencyPrecompile,
+		NFTPrecompile,
+		EVMPrecompile,
+	)>,
 );
 
-impl<PrecompileCallerFilter, MultiCurrencyPrecompile, NFTPrecompile> Precompiles
-	for AllPrecompiles<PrecompileCallerFilter, MultiCurrencyPrecompile, NFTPrecompile>
+impl<PrecompileCallerFilter, MultiCurrencyPrecompile, NFTPrecompile, EVMPrecompile> Precompiles
```
Comment (shaopengw) [inline runtime/common/src/precompile/mod.rs:39]:
```suggestion
impl<PrecompileCallerFilter, MultiCurrencyPrecompile, NFTPrecompile, StateRentPrecompile> Precompiles
```

File: runtime/karura/src/lib.rs [hunk 7]
```diff
@@ -1250,6 +1250,7 @@ pub type MultiCurrencyPrecompile =
 	runtime_common::MultiCurrencyPrecompile<AccountId, EvmAddressMapping<Runtime>, Currencies>;
 
 pub type NFTPrecompile = runtime_common::NFTPrecompile<AccountId, EvmAddressMapping<Runtime>, NFT>;
+pub type EVMPrecompile = runtime_common::EVMPrecompile<AccountId, EvmAddressMapping<Runtime>, EVM>;
```
Comment (shaopengw) [inline runtime/karura/src/lib.rs:1253]: Also change the naming for karura.

**Comment by zjb0807** [comment]
Need to deploy state_rent by genesis and integration test.

**Review by xlc — APPROVED** [review]

**Review by shaopengw — APPROVED** [review]

**Pull request merged**